signals/signal_exporter.py

```diff
@@ -0,0 +1,383 @@
+
+        sig = self._sig_builder(event=event, sample=sample, tag=tag, metadata=metadata)
+        sub_check, result_box = self._check_subs(event)
+
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+        counter = 0
+        while sub_check and sub_check.is_alive():
+            time.sleep(0.1)
+            counter += 1
+            if counter >= 200:
+                print("Timeout after waiting 20 seconds for sub response")
+                return 2
+
+        return result_box[0]
+
+    def initialize(
+        self, legal_events: list, tag: str = None, expected_hosts: list = None
+    ) -> None:
+        """
+        Publishes an initialization message. Starts a listener that reads responses
+        to the initialization message and adds responders to the subscriber list.
+        Sets list of legal event names for future signals, and also allows for optional
+        input of expected hostnames (subscribers) as well as a tag.
+        """
+        if not self._valid_str_list(legal_events):
+            print("ERROR: 'legal_events' arg must be a list of string event names")
+            return
+
+        if expected_hosts:
+            if not self._valid_str_list(expected_hosts):
+                print("ERROR: 'expected_hosts' arg must be a list of string hostnames")
+                return
+            for host in expected_hosts:
+                self.subs.append(host + "-resp")
+
+        self.legal_events = legal_events
+        sig = self._sig_builder(event="initialization", tag=tag)
+        self._fetch_responders()
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+    def shutdown(self, tag: str = None) -> None:
+        """
+        Shuts down initialization response listener (stops accepting subscribers).
+        Wipes the subscriber list and publishes a shutdown message.
+        """
+        sig = self._sig_builder(event="shutdown", tag=tag)
+        self.init_listener.stop()
+        self.subs = []
+        self.redis.publish(channel="event-signal-pubsub", message=sig.to_json_str())
+
+
+class SignalResponder:
+    """
+    A signal management object for tools that wish to respond to event/state signals.
+    Can be used both for listening for signals as well as responding to them. Also
+    allows for locking onto specific tags/publisher_ids. Uses the standard signal
+    response protocol for all published messages.
+    """
+
+    def __init__(self, redis_host: str = "localhost", redis_port: int = 6379) -> None:
+        """
+        Sets exporter object fields and generates unique tool_id.
+        Allows for specification of redis host/port.
+        """
+        self.redis = redis.Redis(host=redis_host, port=redis_port, db=0)
+        self.subscriber = self.redis.pubsub(ignore_subscribe_messages=True)
+        self.subscriber.subscribe("event-signal-pubsub")
+        self.tool_id = platform.node() + "-resp"  # ADD UUID HERE?
```

Comment:
Can we allow for a custom string to be added in here to help with the debugging process? For instance, if I'm running two tools on the same node, I can't differentiate between them. Can we also add in the uuid before merge? If I'm running two tools on the same node, there will be a collision.